### ISSUE TITLE
Undoing a mesh edition and trying to go into edit mode again throws an error and corrupts the mesh

### ISSUE THREAD

**Issue by Gorgious56** [IfcOpenShell/IfcOpenShell#3288]
Steps to reproduce :
- Fresh file
- Add a cube mesh
- Set it to IfcActuator
- Go into edit mode with tab, mess with the geometry
- Go out of edit mode, save changes
- Press Tab, Edit mesh again, press Tab (step not always necessary for bug)
- CTRL - Z a bunch of times to get back to just before modifying the mesh
- Tab into edit mode to edit the mesh
- Error

![mesh_bug](https://github.com/IfcOpenShell/IfcOpenShell/assets/25156105/403e5e65-5920-4a0d-bd36-dc41f2e84ae8)

```
Error: Python: Traceback (most recent call last):
  File "C:\Users\NH\AppData\Roaming\Blender Foundation\Blender\3.5\scripts\addons\blenderbim\bim\module\geometry\operator.py", line 865, in invoke
    return self.execute(context)
  File "C:\Users\NH\AppData\Roaming\Blender Foundation\Blender\3.5\scripts\addons\blenderbim\bim\module\geometry\operator.py", line 759, in execute
    return IfcStore.execute_ifc_operator(self, context)
  File "C:\Users\NH\AppData\Roaming\Blender Fou
ndation\Blender\3.5\scripts\addons\blenderbim\bim\ifc.py", line 410, in execute_ifc_operator
    result = getattr(operator, "_execute")(context)
  File "C:\Users\NH\AppData\Roaming\Blender Foundation\Blender\3.5\scripts\addons\blenderbim\bim\module\geometry\operator.py", line 793, in _execute
    representation = tool.Geometry.get_active_representation(obj)
  File "C:\Users\NH\AppData\Roaming\Blender Foundation\Blender\3.5\scripts\addons\blenderbim\tool\geometry.py", line 132, in get_active_representation
    return tool.Ifc.get().by_id(obj.data.BIMMeshProperties.ifc_definition_id)
  File "C:\Program Files (x86)\Steam\steamapps\common\Blender\3.5\python\lib\site-packages\ifcopenshell\file.py", line 326, in by_id
    return self[id]
  File "C:\Program Files (x86)\Steam\steamapps\common\Blender\3.5\python\lib\site-packages\ifcopenshell\file.py", line 314, in __getitem__
    return entity_instance(self.wrapped_data.by_id(key), self)
  File "C:\Program Files (x86)\Steam\steamapps\common\Blender\3.5\python\lib\site-packages\ifcopenshell\ifcopenshell_wrapper.py", line 4515, in by_id
    return _ifcopenshell_wrapper.file_by_id(self, id)
RuntimeError: Instance #277 not found
```

I don't think it corrupts the ifc file, but it doesn't register mesh changes properly.

ifc file : 
[untitled.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/11715161/untitled.ifc.txt)

**Comment by Moult** [comment]
OK this took a bit of digging. Let me try to explain what I've discovered.

The BBIM add-on undo system works by wrapping all BBIM operator executions (and invokes too). The wrapper begins with generating a random UUID for that operation to identify it, and after the operator finishes, the wrapper ends with storing that operation's UUID in a Blender scene property we call `last_transaction`.

Now when we press undo, an `undo_post` handler checks whether or not the Blender scene property `last_transaction` changes. If it changes, that means that we just undid a BBIM operation (https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/blenderbim/bim/handler.py#L210), and therefore need to also undo potential IFC changes.

If the user runs any other regular vanilla Blender operation, it is not wrapped, and so there is no `last_transaction` change, and therefore we can safely ignore those types of undos.

Now here's where it breaks.

Apparently, if your operator toggles edit mode in addition to setting that scene property, Blender doesn't actually record any state where the scene property has that value. Sorta. Kinda. Take a look at this example.

There is a single scene string prop called `foo`. There are two operators. One operator sets `foo` to a random UUID. Another operator does the same ... but also toggles edit mode. There is an undo_post handler to show you what `foo` is after an undo is done.

Try run the code, then press the `Execute Operator` button a few times, then press undo a few times. As you expect, you see the UUIDs go in reverse.

Now instead press the `Toggle Operator` first then the `Execute Operator`, then press undo twice. Tada! Strangely, now the UUID from the `Toggle Operator` is never restored during the undo. Is this a Blender bug? Why?

I think I can fix this, but I'd love to learn what on earth is going on.

```python
import bpy
import uuid

def update_foo(self, context):
    print("foo updated to:", self.foo)

class SimpleOperatorWithEditToggle(bpy.types.Operator):
    bl_idname = "object.simple_operator_with_edit_toggle"
    bl_label = "Toggle Operator"
    bl_options = {"REGISTER", "UNDO"}


    def execute(self, context):
        context.scene.my_custom_props.foo = str(uuid.uuid4())
        bpy.ops.object.mode_set(mode="EDIT", toggle=True)
        return {'FINISHED'}

class SimpleOperatorWithExecute(bpy.types.Operator):
    bl_idname = "object.simple_operator_with_execute"
    bl_label = "Execute Operator"
    bl_options = {"REGISTER", "UNDO"}


    def execute(self, context):
        context.scene.my_custom_props.foo = str(uuid.uuid4())
        return {'FINISHED'}

class MyProperties(bpy.types.PropertyGroup):
    foo: bpy.props.StringProperty(name="Foo", default="Not set", update=update_foo)

class MyPanel(bpy.types.Panel):
    bl_idname = "OBJECT_PT_my_panel"
    bl_label = "My Custom Addon"
    bl_space_type = 'VIEW_3D'
    bl_region_type = 'UI'
    bl_category = "My Custom Addon"

    def draw(self, context):
        layout = self.layout
        layout.prop(context.scene.my_custom_props, "foo")
        layout.operator('object.simple_operator_with_edit_toggle')
        layout.operator('object.simple_operator_with_execute')

def handle_undo_post(scene):
    print("After Undo:", scene.my_custom_props.foo)

def register():
    bpy.utils.register_class(MyProperties)
    bpy.types.Scene.my_custom_props = bpy.props.PointerProperty(type=MyProperties)
    bpy.utils.register_class(SimpleOperatorWithEditToggle)
    bpy.utils.register_class(SimpleOperatorWithExecute)
    bpy.utils.register_class(MyPanel)
    bpy.app.handlers.undo_post.append(handle_undo_post)

def unregister():
    bpy.utils.unregister_class(MyProperties)
    del bpy.types.Scene.my_custom_props
    bpy.utils.unregister_class(SimpleOperatorWithEditToggle)
    bpy.utils.unregister_class(SimpleOperatorWithExecute)
    bpy.utils.unregister_class(MyPanel)
    bpy.app.handlers.undo_post.remove(handle_undo_post)

if __name__ == "__main__":
    register()
```

**Comment by Andrej730** [comment]
I think I've found it - _"That is known limitation of our undo system, in Edit mode do not tweak Object properties if you want to be able to undo them… Afraid there is nothing else to do here, thanks for the report anyway."_  ([blender bug report](https://projects.blender.org/blender/blender/issues/57897))

I've asked a [question](https://blenderartists.org/t/blender-doesnt-undo-properties-in-edit-mode-looking-for-someone-with-blender-code-knowledge/1473614) on Blender forum - maybe there is some hacky workaround.

There is an option to store data in bmesh but that seems unreliable since there is no way to store data for the entire bmesh, only for verts/edges/faces/loops.

>I think I can fix this, but I'd love to learn what on earth is going on.

you mean by storing `last_transaction` elsewhere besides blender props?

**Comment by Andrej730** [comment]
I guess there are still bunch of bugs if you do any props related action and then undo it while being in EDIT mode.

**Comment by Moult** [comment]
Great find @Andrej730 ! That explains it. Also a huge thanks @Gorgious56 for isolating this case since I'm almost certain users have hit this type of undo bug a lot but simply no one was able to report it meaningfully. Assuming my fix didn't make things worse, this could be a critical fix!

> you mean by storing last_transaction elsewhere besides blender props (just in some python module)?

We already store `last_transaction` in the IfcStore python module to compare whether what we think the last transaction is is the same as what Blender thinks it is... if it isn't, then we have to undo / redo until they get back in sync : https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/blenderbim/bim/handler.py#L210

The fix is that instead of undo / redo one operation at a time, now we undo / redo _to a specific point in history_. So previously the logic was:

 1. If the last transaction key != the transaction key Blender think it is...
 2. ... undo one transaction.

Now the logic is:

 1. If the last transaction key != the transaction key Blender think it is...
 2. Keep on undoing until you hit the transaction with the key.

This means that if Blender "skips" an undo state such as when you toggle edit mode (and in edit mode, the state is ambiguous anyway as the shape is undergoing a change) it doesn't matter as so long as you undo back into an object mode state at some point, the undo system will catch up.

This also means you can use the Edit > Undo History menu now and jump to any point in the history or future. Cool.

Just want to quote this from the upstream Blender bug report. It sounds really important and I don't know if we touch any properties in any other operators which deal with edit mode toggling:

> That is known limitation of our undo system, in Edit mode do not tweak Object properties if you want to be able to undo them… Afraid there is nothing else to do here, thanks for the report anyway.

**Comment by Moult** [comment]
> I guess there are still bunch of bugs if you do any props related action and then undo it while being in EDIT mode.

Exactly. Here's an easy bug you can recreate now.

 1. Create a blank IFC project.
 2. Create a regular Blender cube (no need to assign class)
 3. In Project Setup > Groups press the Load Groups button then add an IFC group.
 4. Undo. Yay! The Group disappears.
 5. Redo. Yay! The group appears again!
 6. Undo. Yay! The group disappears again! Everything works.

... compared to this:

 1. Create a blank IFC project.
 2. Create a regular Blender cube (no need to assign class)
 3. This time, press Tab to enter edit mode for the cube.
 4. In Project Setup > Groups press the Load Groups button then add an IFC group.
 5. Undo. Huh? The group is still there?
 6. Undo again. Huh? Still there?
 7. Undo again. Huh? Now the cube is in object mode, but the group is still there?
 8. Undo again. Huh? Now the cube gets deleted aaaaaand .... oh! The group has finally disappeared (i.e. our undo system has caught up).

But it's not _that_ bad. For example, if instead this time after step 4 you pressed Tab again to go back into Object mode, you'll see that the undo system "works" again (i.e. state is "as expected" before and after editing).

**tl;dr if you choose to edit your IFC data (e.g. edit a pset) whilst in Edit Mode, always go back to Object Mode before pressing undo or you might lose data. To be safe, only edit geometry when in Edit Mode.**

**Comment by Moult** [comment]
Ahh thanks @Andrej730 just realised you have video evidence of this bug and its behaviour in vanilla Blender and responses on BA: https://blenderartists.org/t/blender-doesnt-undo-properties-in-edit-mode-looking-for-someone-with-blender-code-knowledge/1473614/8

Ping @theoryshaw @chubbyquark @Plae-blue-dot thought you guys might like to be aware of this Blender issue.

**Comment by Andrej730** [comment]
Maybe we should keep the issue open until find some UX solution to this? 

Basically the current best practice for this is - do your actions that are *requiring* EDIT mode and then get out of it to make sure nothing will go wrong.
And note to us as developers to make sure operators that do stuff in EDIT mode wouldn't break on undo given that all blender properties wouldn't undo.

~~would be nice to find some other more critical Blender bug that would encourage developers to fix this~~

**Comment by Moult** [comment]
Well, here's the worst case scenario: The total amount of data that could be lost is the amount of data edited 1) whilst the user is currently in edit mode and 2) if the user then presses undo enough times to go back into object mode.

In practice, I can't imagine how much data that would be. (but then again I'm the type of user who never ever presses undo and so never encounters a single undo bug muahahaha)

A really extreme UX solution is to have panels poll for mode and if in edit mode literally disable all edit / save / add / remove buttons. No thanks.

Maybe another way is to inject a warning into the header, footer, status bar or something that if in edit mode shows a caution icon and some sort of message but I have no idea what short message would communicate the nuances of this situation to a user and not just scare them into thinking they're doing something wrong.

I'm fine to leave it be personally esp. as this is fundamental upstream behaviour. Curious what others think.

**Comment by theoryshaw** [comment]
> Also a huge thanks @Gorgious56 for isolating this case since I'm almost certain users have hit this type of undo bug a lot but simply no one was able to report it meaningfully.

Yes, i always got nervous when i undo'ed too many times...as errors would pop up.  :)  Couldn't narrow in on the exact problem, however.

**Comment by Gorgious56** [comment]
This is AWESOME ! I'll play with that and test the limits for sure. I'm a CTRL-Z'er for life ! :) Depends on how much this issue pops up, but maybe an info message could pop up if the user tries to modify a property in edit mode ? I don't think disabling properties is reasonable though, because it is still a valid workflow, it just makes the undo system finicky to work with.
